### ISSUE TITLE
variable-seed-test is stochastic

### ISSUE THREAD

**Issue by joinr** [incanter/incanter#385]
variable-seed-test in incanter.stats-tests is sensitive to how (Date.) is providing a seed for the prng for sample-uniform.  The assumption in the test is that, providing a seed = deterministic sampling, no seed provided = random sampling, thus the samples should be different if seed options are different.  The second seed option is provided by a destructured optional argument for seed, (Date.).  I'm getting periodic test fails from this particular test.  Not sure what the structural details are (maybe my system timer is too fast, or not, dunno).

**Comment by MawiraIke** [comment]
This issue seems to be fixed? I have run the variable-seed-test a couple of times and its passing. I can't guess why it would fail at the current state, i see the seed is passed to the MersenneTwister well.

**Comment by joinr** [comment]
>I have run the variable-seed-test a couple of times and its passing.

I haven't tried this out in a while.  I kind of stopped messing with Incanter after I got stuck trying to "fix" the incanter.charts implementation by wrapping a better version of jfreechart.  It could very well be fixed, although the original problem failed to manifest several times (e.g. my remark about the stochastic nature).

**Comment by MawiraIke** [comment]
Were you trying to move to v1.5.0? From the migration guide, at their [GitHub](https://github.com/jfree/jfreechart) page, it seems like migrating from the current v1.0.19 is a daunting task with a lot of renaming and import changes.

**Comment by joinr** [comment]
No.  I was actually fixing 1.9 to be compatible with 1.58 so that the branches could be merged into a single dev branch instead of the weird mismatched versions.  1.9 was kind of a disaster organizationally, in that is migrated a ton of stuff to clojure.core.matrix without really 2x checking to maintain backwards compatibility.  Consequently, many bugs emerged after the fact, and users (like me) stuck with 1.58 for a long time.  I resolved to end this and get a single branch working, about a year ago.  I put in a bunch of fixes, merges, and patches, which eventually were accepted.  During this time, I was running the regression tests and didn't have problems - until I did.  Again, nothing changed in the relevant code base killing the tests.  This led me to ID the bug as possibly stochastic in nature (despite outward appearances and many successful test runs).

Note: Incanter is pretty much on life support.  It's got some good ideas, but the libraries are outdated, the API is kind of frozen (a lot of implementation details reach back to 2010), some of the functions are subpar (xls parsing for instance, modeling some distributions), etc.  I had planned to revamp it on my branch and push changes upstream, but I don't know if that will pan out.

**Comment by MawiraIke** [comment]
Oh. Thank you for your reply. I will check the function again, hope i will fish out something.

**Comment by MawiraIke** [comment]
@joinr Does Incanter provide any 3D charts? Can developers contribute them?
I would like to work on some on my fork but don't see any existing 3D charts so i am not sure of the policy.

**Comment by joinr** [comment]
Incanter charts are tied to an old version if jfreechart, and do not leverage all available plots.  Jfreechart had a big change in recent versions and altered it's (pretty complicated) object oriented api.  I think there are 3d plots there, but you have to wrap them.  As far as contributing,  I think contributions are welcome, but there is high latency in getting things accepted.  It may be better to fork incanter, add changes for your own use, and submit downstream patches.

**Comment by MawiraIke** [comment]
I am not using incanter for personal use, I recently started using Clojure and was contributing to incanter to add up my skills in Clojure. I noticed about the latency in merging PR's into Incanter and was hoping to contribute what I can and have them merged at a later time, hopefully. 
About the old JFreeChart issue, is there anything that can be done to replace it with the new 1.5.0? I assume there will be a lot of code breaking changes.

**Comment by joinr** [comment]
See #389 .  I did some initial exploration at differences and started writing a more complete wrapper for jfreechart 1.5, but honestly got bogged down trying to map out all the weird combinations of dataset types and chart types.

**Comment by MawiraIke** [comment]
I will try the migration to jfreechart 1.5 on my fork and see how it goes. Hope it works, if it does I will submit a PR to this repo and wait for the merge. If it isn't merged, or it takes time, I hope it will save someone else some time.

**Comment by joinr** [comment]
The [cljfreechart](https://github.com/joinr/cljfreechart) lib has some of this mapped out.  In particular, the xyplot.graphml file is a mapping of all the internal JFreeChart data series to plot types, and a mapping of intended clojure protocols and incanter.chart types.  I started reorganizing the legacy (janky) multimethod-based approach into a set of protocols, as well as extending the coverage of plot types to include all of what jfreechart offers (incanter only had a subset, e.g. stacked area plots were strickly categorical since original author didn't include continuous x-axis plots from jfree).

I'd also refer to my notes from #389, which maps out the minimal changes necessary to get legacy incanter.charts working with jfree 1.5 (in light of deprecations).  I think, longer term, it's more desirable to either have a good wrapper around jfreechart, or even better, use a platform like vega (e.g. [vegafx](https://github.com/joinr/vegafx) ) or [cljplot](https://github.com/generateme/cljplot).  If you spend time with JFreeChart, you'll quickly find out how java-centric it is....lots of classes, lots of factories.  It's got some degree of completeness in this respect though, although other aforementioned plotting libraries are capable of "more" IMO.

**Comment by MawiraIke** [comment]
I like the idea of creating a wrapper around JFreeChart, I had started adding multiple charts that incanter had left out, like Polar Charts, Ring Charts and had others in mind like the 3D charts but soon realized that my method would take forever to implement all the capabilities of JFreeChart. The idea of a wrapper would suit this problem space greatly. Else, as you suggested, I think a platform like vega would be better as it provides neater charts with more features like sharing and side by side code and chart viewing. I found this repo, [oz](https://github.com/metasoarous/oz) that does something similar although it doesn't implement jfree and is for both clj and cljs.

I have checked the sci cloj repos and really liked the work there, looking forward to making my contributions there. I will go through the cljfreechart repo you sent and hopefully contribute there too.
Thank you for your timely replies and for the links.
